pkg/virtctl/imageupload/imageupload.go

```diff
@@ -0,0 +1,304 @@
+}
+
+func (c *command) run(cmd *cobra.Command, args []string) error {
+	file, err := os.Open(imagePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	namespace, _, err := c.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(c.clientConfig)
+	if err != nil {
+		return fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
+	}
+
+	pvc, err := getUploadPVC(virtClient, namespace, pvcName, noCreate)
+	if err != nil {
+		return err
+	}
+
+	if pvc == nil {
+		pvc, err = createUploadPVC(virtClient, namespace, pvcName, pvcSize, storageClass, accessMode)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("PVC %s/%s created\n", namespace, pvc.Name)
+	} else {
+		fmt.Printf("Using existing PVC %s/%s\n", namespace, pvc.Name)
+	}
+
+	err = waitUploadPodRunning(virtClient, namespace, pvcName, uploadPodWaitInterval, uploadPodWaitTimeout)
+	if err != nil {
+		return err
+	}
+
+	token, err := getUploadToken(virtClient.CdiClient(), namespace, pvcName)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Uploading data to %s\n", uploadProxyURL)
+
+	err = uploadData(uploadProxyURL, token, file, insecure)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Uploading %s completed successfully\n", imagePath)
+
+	return nil
+}
+
+func getHTTPClient(insecure bool) *http.Client {
+	client := &http.Client{}
+
+	if insecure {
+		client.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+
+	return client
+}
+
+func uploadData(uploadProxyURL, token string, reader io.Reader, insecure bool) error {
+	url := uploadProxyURL + uploadProxyURI
+	req, _ := http.NewRequest("POST", url, reader)
+	client := httpClientCreatorFunc(insecure)
+
+	req.Header.Add("Authorization", "Bearer "+token)
+	req.Header.Add("Content-Type", "application/octet-stream")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("Unexpected return value %d", resp.StatusCode)
+	}
+
+	return nil
+}
+
+func getUploadToken(client cdiClientset.Interface, namespace, name string) (string, error) {
+	request := &uploadcdiv1.UploadTokenRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "token-for-virtctl",
+		},
+		Spec: uploadcdiv1.UploadTokenRequestSpec{
+			PvcName: name,
+		},
+	}
+
+	response, err := client.UploadV1alpha1().UploadTokenRequests(namespace).Create(request)
+	if err != nil {
+		return "", err
+	}
+
+	return response.Status.Token, nil
+}
+
+func waitUploadPodRunning(client kubernetes.Interface, namespace, name string, interval, timeout time.Duration) error {
+	loggedStatus := false
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		pvc, err := client.CoreV1().PersistentVolumeClaims(namespace).Get(name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		value, _ := pvc.Annotations[podPhaseAnnotation]
+
+		done := (value == string(v1.PodRunning))
```

Comment:
Is there any reason to bail early if the pod status is PodFailed?  Would that ever happen here?

Reply:
Upload pods are restarted on failure.